srt-protocol/src/crypto.rs

```diff
@@ -270,6 +270,68 @@ impl CryptoManager {
 
         ret
     }
+
+    pub fn rekey(&mut self, kmreq: &SrtKeyMessage) -> Result<(), ConnectionReject> {
+        assert_eq!(kmreq.pt, PacketType::KeyingMaterial);
+
+        let salt = kmreq.salt[..].try_into().unwrap();
+        let kek = CryptoManager::gen_kek(&self.options, &salt);
+
+        assert_eq!(
+            kmreq.wrapped_keys.len(),
+            kmreq.key_flags.bits().count_ones() as usize * usize::from(self.options.size) + 8
+        );
+
+        let mut keys = vec![0; kmreq.wrapped_keys.len() - 8];
+
+        let mut iv = [0; 8];
+        match kek.len() {
+            16 => wrap::aes_unwrap(
+                &Aes128::new(kek[..].into()),
+                &mut iv,
+                &mut keys,
+                &kmreq.wrapped_keys,
+            ),
+            24 => wrap::aes_unwrap(
+                &Aes192::new(kek[..].into()),
+                &mut iv,
+                &mut keys,
+                &kmreq.wrapped_keys,
+            ),
+            32 => wrap::aes_unwrap(
+                &Aes256::new(kek[..].into()),
+                &mut iv,
+                &mut keys,
+                &kmreq.wrapped_keys,
+            ),
+            _ => panic!("Invalid key size"),
+        }
+
+        if iv != wrap::DEFAULT_IV {
+            return Err(ConnectionReject::Rejecting(
+                CoreRejectReason::BadSecret.into(),
+            ));
+        }
+
+        let even = if kmreq.key_flags.contains(KeyFlags::EVEN) {
+            Some(keys[0..usize::from(self.options.size)].into())
+        } else {
+            None
+        };
+        let odd = if kmreq.key_flags.contains(KeyFlags::ODD) {
+            Some((keys[keys.len() - usize::from(self.options.size)..]).into())
+        } else {
+            None
+        };
+
+        self.salt = salt;
+        self.odd_sek = odd;
+        self.even_sek = even;
```

Comment:
If the recieved message didn't contain a given key, should we actually leave the existing key in place rather than overwriting with `None`?

Reply:
Looks like [hcrypt_ctx_rx.c](https://github.com/Haivision/srt/blob/257e022337cc6e15239663d34b0d8fe2a6d61ac0/haicrypt/hcrypt_ctx_rx.c) might be the relevant section of the reference implementation, but I've not understood it yet.

Reply:
This is suspicious to me--this would discard any keys we already have, which I would think would result in a panic immediately.... (it shoudn't, seems like that's a security problem rn, anyone who sends an encrypted packet just will panic if you have the wrong key set!!!)

You should only assign to each sek if the tag is in the bitfield, ie, get rid of these two lines and just assign directly in the if statement.